src/mongoc/mongoc-cursor.c

```diff
@@ -937,10 +935,19 @@ _mongoc_cursor_get_host (mongoc_cursor_t    *cursor,
       return;
    }
 
-   *host = cursor->client->cluster.nodes[cursor->hint - 1].host;
-   host->next = NULL;
-}
+   description = mongoc_topology_server_by_id(cursor->client->topology, cursor->hint);
+   if (!description) {
+      MONGOC_WARNING("%s(): Invalid cursor hint, no matching host.",
+                     __FUNCTION__);
+      return;
+   }
 
+   *host = description->host;
```

Comment:
this is pretty minor, but you should null out host.next so you don't leave it dangling.